Adding Int128 and UInt128 with a base software implementation

tannergooding opened the pull request:
This does not cover acceleration in the JIT, that will be a separate work item and will be handled in a follow up PR.

tannergooding committed:
Adding barebones Int128 and UInt128 structs

tannergooding committed:
Special case Int128 and UInt128 alignment on x64 Unix and Arm64

tannergooding committed:
Implementing Int128 and UInt128

tannergooding committed:
Adding tests for Int128 and UInt128

tannergooding committed:
Updating Int128/UInt128 to respect the System V ABI ordering

Perksey commented on src/libraries/System.Private.CoreLib/src/System/UInt128.cs:
```diff
@@ -0,0 +1,1654 @@
+            // of 32-bit unsigned integers described
+            // in Hacker's Delight by Henry S. Warren, Jr. (ISBN 0-201-91465-4), Chapter 8
+            // Basically, it's an optimized version of FOIL method applied to
+            // low and high dwords of each operand
```
Copy + pasted comment, it’s not dwords anymore ;)

tannergooding replied:
Fixed.

jkotas commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -9898,6 +9897,21 @@ void MethodTableBuilder::CheckForSystemTypes()
+
+                if ((strcmp(name, g_Int128Name) == 0) || (strcmp(name, g_UInt128Name) == 0))
+                {
+                    pLayout->m_ManagedLargestAlignmentRequirementOfAllMembers = 16; // sizeof(__int128)
```
This needs equivalent fix in crossgen/NativeAOT and in Mono

tannergooding replied:
@fanyang-mono could you point me to where `Mono` handles custom struct packing/alignment?

I tried checking for where that's handled for the vector types, but couldn't find it.

`Int128` and `UInt128` need to have 16-byte packing when targeting the `System V` ABI (used by Unix) and the same packing/alignment as `Int64`/`UInt64` otherwise (such as on Windows or 32-bit platforms).

tannergooding replied:
-- I fixed this up for crossgen/NativeAOT already and added corresponding P/Invoke tests to ensure the data is passed correctly to/from Native.

vargaz replied:
It's handled in mono_class_layout_fields () in metadata/class-init.c.
Not sure the mono gc supports 16 byte aligned objects on 32 bit platforms.

tannergooding replied:
This doesn't impact 32-bit platforms, only 64-bit platforms (and only 64-bit Unix at that).

tannergooding replied:
@vargaz, looks like its just blocked everywhere not just 32-bit platforms: https://github.com/dotnet/runtime/blob/main/src/mono/mono/metadata/class-init.c#L2058-L2065

It looks like Mono also isn't differentiating "alignment" from "packing" and so the layout of types that include the new `Int128`, `UInt128`, or the existing `Vector64<T>`, `Vector128<T>`, `Vector256<T>` types will be incorrect.

This would explain why I couldn't find any logic touching `Vector128<T>` and ensuring that it has the right packing (and therefore layout) even if the GC couldn't correctly align the overall allocation.

vargaz replied:
Yes, it looks like this is not implemented right now in mono.

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into generic-math-int128

tannergooding committed:
Fixing an issue with UInt128->BigInteger setting the wrong sign

tannergooding commented:
Hmmm. Tests were all passing locally but failing in CI, going to need to investigate this more

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into generic-math-int128

tannergooding committed:
Don't use Unsafe.As in the Int128/UInt128 hex parsing logic

tannergooding committed:
Don't use Unsafe.As in the Int128/UInt128 hex parsing logic

tannergooding commented:
This should just need the NativeAOT and Mono fixups to ensure `Int128` and `UInt128` are `Pack = 16` on Sys V (x64 and Arm64)

tannergooding commented:
I also apparently forgot to implement conversions to and from `float`/`double`/`decimal`, so I will do that and add tests

tannergooding committed:
Adding Int128 P/Invoke tests and ensure R2R correctly sets the packing

tannergooding committed:
Fixing some issues with the Int128 interop test for non-Windows

tannergooding committed:
Ensure that floating-point conversions exist for Int128 and UInt128

deeprobin commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilerContext.cs:
```diff
@@ -36,6 +36,7 @@ public partial class ReadyToRunCompilerContext : CompilerTypeSystemContext
         private SystemObjectFieldLayoutAlgorithm _systemObjectFieldLayoutAlgorithm;
         private VectorOfTFieldLayoutAlgorithm _vectorOfTFieldLayoutAlgorithm;
         private VectorFieldLayoutAlgorithm _vectorFieldLayoutAlgorithm;
+        private Int128FieldLayoutAlgorithm _Int128FieldLayoutAlgorithm;
```
```suggestion
        private Int128FieldLayoutAlgorithm _int128FieldLayoutAlgorithm;
```
I think you should keep the naming consistent.

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        }
+
+        /// <inheritdoc cref="object.Equals(object?)" />
+        public override bool Equals([NotNullWhen(true)] object? obj)
```
Wouldn't a check with `GetType()` make more sense here?
See https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aaGfef6m5JA2AQIAGwYBRAI4BXbONwBBABQRgAKxhgMDDZoCU3Pr07VTlpgHYG6rQwCWuBuIgA7AOb6MACxhQhgDcJpYAvqGmAsRCImKSsgpKAEL22rr6WsYWVuZWUbYGAHQA4jAYACoAngAOMKqGDAC8TQwYtTAQAGaqbl7BkXwR1GFAA===

tannergooding replied:
No. The `obj` must be unboxed if the type is `Int128` and a comparison must then be done using the unboxed value.

`x is T other` is the standard pattern for that and the runtime should optimize it accordingly.

deeprobin commented on THIRD-PARTY-NOTICES.TXT:
```diff
@@ -1072,4 +1072,30 @@ Copyright (c) Microsoft Corporation.
\ No newline at end of file
+https://github.com/microsoft/msquic/blob/main/LICENSE
+
+License notice for m-ou-se/floatconv
```
Where did you use code licensed for `m-ou-se/floatconv`?

tannergooding replied:
In https://github.com/dotnet/runtime/pull/69204/files#diff-68eb1454c605f9d2192816343fd055ac5482a7b001e1f010f39c9ab1777ad441R252 (and another place) where the corresponding same callouts exist.

deeprobin reviewed (approved):
Have noted a few minor things.
But looks very good. 

Looking forward to `Int128` and `UInt128` 🎉😄.

tannergooding committed:
Fixing the casing of a couple fields

tannergooding committed:
Revert "Don't use Unsafe.As in the Int128/UInt128 hex parsing logic"

This reverts commit 09e8bfc52d5b1b7f7b341f318bf1046b19768dd3.

tannergooding committed:
Adjusting the Int128/UInt128 generic math tests to have consistent ordering

dakersnar commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        internal ulong Upper => _upper;
+
+        /// <inheritdoc cref="IComparable.CompareTo(object)" />
+        public int CompareTo(object? value)
```
Can we not compare across types? Is that not supported?

dakersnar reviewed (approved):
Looks good! Left some comments, some of which are suggestions, others are clarification questions. I checked Int128.cs and UInt128.cs pretty thoroughly. The formatting and parsing code I did not examine closely. I skimmed the test files; I'm assuming they are pretty formulaic and therefore exhaustive. If you address my small comments and are confident with the formatting, parsing, and tests, I think this is ready to merge.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+            }
+            else
+            {
+                throw new ArgumentException(SR.Arg_MustBeInt128);
```
Right. `CompareTo` traditionally only allows comparisons to the "same type". It's documented to throw if the type isn't supported.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        }
+
+        /// <inheritdoc cref="object.Equals(object?)" />
+        public override bool Equals([NotNullWhen(true)] object? obj)
```
`Equals` is logically "not equal" when the types are different where-as `CompareTo` can't really report anything. This is the standardized behavior for the two methods.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        /// <inheritdoc cref="object.ToString()" />
+        public override string ToString()
+        {
+            return Number.Int128ToDecStr(this);
```
Its traditionally been centralized to allow logic to be shared between all integer and floating-point types where possible.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        {
+            const double TwoPow127 = 170141183460469231731687303715884105728.0;
+
+            if ((value < -TwoPow127) || double.IsNaN(value) || (value >= +TwoPow127))
```
Yes.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+                ThrowHelper.ThrowOverflowException();
+            }
+
+            if (~value._upper == 0)
```
This looks like a copy paste error. This is meant to be ordered:
```csharp
if (~value._upper == 0)
{
    long lower = (long)value._lower;
    return checked((short)lower);
}


if (value._upper != 0)
{
    ThrowHelper.ThrowOverflowException();
}
```

Basically for checked `Int128->smaller unsigned` conversion, `value._upper` must be `0`.

For checked `Int128->smaller signed` conversion, `value._upper` must be `0` (positive) or `~0` (all bits set, representing a negative value that has 64 or fewer bits)

tannergooding replied:
Fixed.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        }
+
+        /// <inheritdoc cref="IComparisonOperators{TSelf, TOther}.op_GreaterThan(TSelf, TOther)" />
+        public static bool operator >(Int128 left, Int128 right)
```
I'm not confident enough that the JIT will optimize that correctly in all the scenarios and so I'd rather it be explicit here.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+        /// <inheritdoc cref="INumber{TSelf}.IsNegative(TSelf)" />
+        public static bool IsNegative(Int128 value) => (long)value._upper < 0;
+
+        internal static bool IsPositive(Int128 value) => (long)value._upper >= 0;
```
Yes. This is consistent with how `IEEE 754` treats `0` for floating-point (`0` is positive, `-0` is negative).

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/UInt128.cs:
```diff
@@ -0,0 +1,1835 @@
+        {
+            const double TwoPow128 = 340282366920938463463374607431768211456.0;
+
+            if ((value < 0.0) || double.IsNaN(value) || (value >= TwoPow128))
```
This is because of `-0`. Basically for the purposes of `floating-point` to `unsigned integer` conversions, `-0` is treated as `+0` and not as an invalid negative number.

I'll add a comment.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/UInt128.cs:
```diff
@@ -0,0 +1,1835 @@
+                );
+            }
+
+            static uint SubtractDivisor(Span<uint> left, ReadOnlySpan<uint> right, ulong q)
```
I just order them alphabetically because it helps me find them more easily.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+
+            uint sign = (uint)(left._upper >> 63);
+
+            if ((sign ^ (uint)(right._upper >> 63)) == 0)
```
Fixed.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Int128.cs:
```diff
@@ -0,0 +1,1832 @@
+
+            if ((sign ^ (uint)(right._upper >> 63)) == 0)
+            {
+                if ((sign ^ (uint)(result._upper >> 63)) != 0)
```
Fixed.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/UInt128.cs:
```diff
@@ -0,0 +1,1835 @@
+        // Explicit Conversions To UInt128
+        //
+
+        /// <summary>Explicitly converts a <see cref="decimal" /> value to a 128-bit signed integer.</summary>
```
Fixed.

tannergooding committed:
Responding to PR feedback

tannergooding commented:
@MichalStrehovsky, @jkotas could you glance over the VM and NativeAOT changes covering the packing requirements?

jkotas commented on src/coreclr/vm/corelib.h:
```diff
@@ -282,6 +282,9 @@ DEFINE_FIELD(DELEGATE,            METHOD_PTR_AUX,         _methodPtrAux)
 DEFINE_METHOD(DELEGATE,             CONSTRUCT_DELEGATE,     DelegateConstruct,          IM_Obj_IntPtr_RetVoid)
 DEFINE_METHOD(DELEGATE,             GET_INVOKE_METHOD,      GetInvokeMethod,            IM_RetIntPtr)
 
+DEFINE_CLASS(INT128,               System,                 Int128)
```
Is this used anywhere?

tannergooding replied:
This is what defines `g_Int128` which is used in `methodTableBuilder.cpp` (https://github.com/dotnet/runtime/pull/69204/files/cb5a82e5df48bf4a7c6deea85765c75d3d4a2e03#diff-4b65f34bf68fdb80cc734cc2435c72ef7f86ef16b9f56af47c06b87761383d3aR9909)

jkotas replied:
> This is what defines g_Int128

`g_Int128Name` is defined in classnames.h. This macro defines `CLASS_INT128` that I do not see used anywhere.

tannergooding replied:
Ah right, I was getting two two things mixed up.

Looks like this is missing handling in `classlayoutinfo.cpp` which copies the alignment requirement to the `pNativeLayoutInfo`: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/classlayoutinfo.cpp#L947-L964

tannergooding replied:
Should be fixed. The only other usage of `CLASS__*` for the `Vector*` types was blocking interop which isn't required for Int128/UInt128 given the interop tests are all passing.

tannergooding committed:
Ensure that pNativeLayoutInfo alignment is initialized for Int128/UInt128

jkotas commented:
> ibraries Test Run checked coreclr Linux arm Debug
> ibraries Test Run checked coreclr Linux_musl arm64 Debug

Both these runs fails with GC asserts. I do not think we have any issues tracking failures like this. Is it caused by the changes in this PR?

tannergooding commented:
> Both these runs failed with GC asserts. I do not think we have any issues tracking failures like this. Is it caused by the changes in this PR?

I'll take a look but I find it doubtful that it'd be caused by this PR. It wasn't failing before and the only change since the last run was to update the native layout for `Int128/UInt128`. Neither of these types are used by the `Regex` or `Http` tests.

`Http` is failing `BOOL MethodTable::SanityCheck()` for:
```
    // strings have component size2, all other non-arrays should have 0
    _ASSERTE((GetComponentSize() <= 2) || IsArray());
```

The other (`Regex`) is failing `_ASSERTE(RawGetMethodTable())` in `void CObjectHeader::SetMarked()`

dakersnar reviewed (approved):
Changes look good, thanks!

jkotas commented:
> Neither of these types are used by the Regex or Http tests.

These asserts are generic signs of GC heap corruption. A bug introduced in common parts (e.g. parsing/formatting) can lead to intermittent crash like that.

tannergooding commented:
> These asserts are generic signs of GC heap corruption. A bug introduced in common parts (e.g. parsing/formatting) can lead to intermittent crash like that.

I'd speculate there may be another issue with `Unsafe.As` then and there is still something wrong with https://github.com/dotnet/runtime/pull/69204/commits/7666952d06c1280d0f0c8c117d599a76df4c0a8d

There was a similar issue exposed by making these "intrinsic" that was fixed in https://github.com/dotnet/runtime/pull/69271 and its possible this is another case that hasn't been found yet.

tannergooding commented:
It could be related to https://github.com/dotnet/runtime/pull/69501 for example, which @SingleAccretion just opened and impacts structs.

tannergooding commented:
If this one is also related to `Unsafe.As`, then it may be worth reverting just that one from intrinsic handling for the time being, as this will have been the 3rd (iirc) GC related issue it exposed.

tannergooding commented:
I'm not 100% this fixes the issue as its not a reliable repro. But I did hit it in 20 runs with the `Unsafe.As` usage and didn't hit it in 40 without it removed.

jkotas commented:
/azp run runtime-extra-platforms

jakobbotsch commented:
> I'm not 100% this fixes the issue as its not a reliable repro. But I did hit it in 20 runs with the `Unsafe.As` usage and didn't hit it in 40 without it removed.

Can you give instructions for how to repro this? If we have a repro for it we need to look at it, it could be manifesting in other places that are way harder to diagnose.

tannergooding commented:
> Can you give instructions for how to repro this? If we have a repro for it we need to look at it, it could be manifesting in other places that are way harder to diagnose.

It was just running the `Int128`/`UInt128` tests with and without the `Unsafe.As`. I imagine it would reproduce for any non-primitive struct.

jkotas commented:
The new Int128 PInvoke test is failing on Android in the extra platforms leg.

tannergooding commented:
Looks like it can't resolve the native library: `System.DllNotFoundException: Int128Native`

@vargaz is there anything additional required for P/Invoke to work on Android? Does the library have or need some special name that differs from the default for example?

tannergooding commented:
The tvOS and iOS failures look unrelated, they seem to have just failed and look to be doing that in other jobs as well.

tannergooding commented:
Hmmm, it looks like most of the `Interop/PInvoke/*` tests are currently disabled on Mono. See https://github.com/dotnet/runtime/blob/main/src/tests/issues.targets#L1716-L1772 under https://github.com/dotnet/runtime/blob/main/src/tests/issues.targets#L1446-L1447; and more further down and in other mono areas.

Most are marked "needs triage" as well.

tannergooding commented:
There are various other `extra-platforms` failures but the `android` issue is the only one that looks related. Others look to be failures more broadly impacting these jobs in particular, many of them are just general timeouts at 240 minutes.

Given the breadth of P/Invoke tests disabled for Android, I'd be inclined to log an issue tracking Mono enabling interop here. The alternative would be to just block `Int128`/`UInt128` from all interop until `Mono` can add support.

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into generic-math-int128

tannergooding committed:
Skip the Interop/PInvoke/Int128 tests on Mono

tannergooding commented:
Logged https://github.com/dotnet/runtime/issues/69531
